rename the nick/remark/display for contact/room #217

lijiarui opened the pull request.

zixia commented on src/room.ts:
```diff
@@ -25,12 +25,12 @@ type RoomObj = {
-  displayMap: Map<string, string>
+  nameMap:      Map<string, string>
+  aliasMap:     Map<string, string>
+  roomAliasMap: Map<string, string>
```
Remove `roomAliasMap` here. Because `aliasMap` should be `roomAliasMap`

The room should only in charge of search **alias in room**. Search contact alias should use `Contact.find()` instead.

lijiarui replied:
If we use Contact.find() instead, then when use function `member()` to get Contact would be asynchronous， which means it will use 
`public async member(queryArg: MemberQueryFilter | string): Promise<Contact | null>` 
but actually we want to use synchronous function.

And, we must use alias(remark) in room, which I have said in #104 :

You said in [room.ts](https://github.com/wechaty/wechaty/blob/master/src/room.ts)

> ISSUE #104 never use remark name because sys group message will never use that

**But sys group message use remark.** 

For example, when the bot set remark for A, then A changed the topic, then sys message shows the remark, not roomAlias.

I have summarize in [#173](https://github.com/wechaty/wechaty/issues/173):

>

> Display order:
> 
> @ Event
> 
> When someone @ a contact in a room, wechat recognise the name order as follows:
> displayName > nickName
> 
> system message
> 
> When room event happens(join, leave, changetopic), system message recognise the name order as follows:
> remark > nickName
> 
> on-screen Names
> 
> The contact name showed to the wechat user in the group
> remark > displayName > nickName

lijiarui replied:
So，I recommend we should use three kind of map:

> nameMap
> aliasMap
> roomAliasMap

zixia replied:
Ok, you are right. 

Agreed.

zixia commented on src/room.ts:
```diff
@@ -25,12 +25,12 @@ type RoomObj = {
 }
 
-type NameType = 'nick' | 'display' | 'remark'
+type NameType = 'name' | 'alias' | 'roomAlias'
```
no `roomAlias`. `alias` is for room only.

zixia commented on src/room.ts:
```diff
@@ -57,9 +57,9 @@ export type RoomQueryFilter = {
-  display?: string
+  name?:      string
+  alias?:     string
+  roomAlias?: string
```
no roomAlias.

zixia commented on src/room.ts:
```diff
@@ -350,11 +350,20 @@ export class Room extends EventEmitter implements Sayable {
   }
 
+  // should be deprecated
   public nick(contact: Contact): string {
```
`nick` should be renamed to `alias`  for the consistency.

lijiarui commented on src/contact.ts:
```diff
@@ -55,8 +55,10 @@ export enum Gender {
-  name?: string | RegExp
+  name?:   string | RegExp
+  // remark should be deprecated
   remark?: string | RegExp
```
Then we should deprecate function `Contact.find(remark: string) ` right now?

zixia replied:
deprecate right now. but it should still functional.

lijiarui replied:
ok, I guess I can use function `Contact.find(query: ContactQueryFilter | {remark: string})` instead, is it right?

lijiarui commented on src/contact.ts:
```diff
@@ -232,6 +234,10 @@ export class Contact implements Sayable {
     }
 
+    if (query[0] === 'remark') {
+      log.warn('Contact', 'findAll(remark:%s) DEPRECATED, use findAll(alias:%s) instead.')
```
no need to add `query[0] = 'alias'`, because I had deal the alias in the following code:
```
const keyMap = {
      name:   'NickName',
      // should be deprecated
      remark: 'RemarkName',
      alias:  'RemarkName',
    }
...
 if (filterValue instanceof RegExp) {
      filterFunction = `(function (c) { return ${filterValue.toString()}.test(c.${filterKey}) })`
    } else if (typeof filterValue === 'string') {
      filterValue = filterValue.replace(/'/g, '\\\'')
      filterFunction = `(function (c) { return c.${filterKey} === '${filterValue}' })`
    } else {
      throw new Error('unsupport name type')
    }
```

If you are sure to deprecate function `Contact.find(remark:string)` right now, I will delete this

zixia replied:
no, you misunderstand.

you should remove `remark` in keyMap first, then you will have to follow my request.

lijiarui commented on src/contact.ts:
```diff
@@ -287,46 +295,57 @@ export class Contact implements Sayable {
                   })
   }
 
+  public remark(newRemark?: string|null): Promise<boolean> | string | null {
```
should  `log.warn('Contact', 'remark(%s) DEPRECATED, use alias(%s) instead.')` be ok? I have added it under the function

zixia commented on src/contact.ts:
```diff
@@ -287,51 +292,63 @@ export class Contact implements Sayable {
+  public static async find(query: ContactQueryFilter | {remark: string | RegExp}): Promise<Contact> {
+    log.verbose('Contact', 'find(%s)', query)
 
     const contactList = await Contact.findAll(query)
```
should also `log.warn()` for `remark` in `find()` function.

lijiarui replied:
Well, I didn't add log warn  because `find()` use `findAll()`, and I had add `log.warn()` for  `findAll()`. 

Do you agree?

zixia commented on src/room.ts:
```diff
@@ -350,11 +351,17 @@ export class Room extends EventEmitter implements Sayable {
+  // should be deprecated
   public nick(contact: Contact): string {
-    if (!this.obj || !this.obj.nickMap) {
+    log.warn('Room', 'nick(Contact) DEPRECATED, use name(Contact) instead.')
```
`name(Contact)` here should be `alias(Contact)`?

zixia commented on src/room.ts:
```diff
@@ -399,7 +406,7 @@ export class Room extends EventEmitter implements Sayable {
   public member(queryArg: MemberQueryFilter | string): Contact | null {
     if (typeof queryArg === 'string') {
-      return this.member({remark: queryArg}) || this.member({display: queryArg}) || this.member({nick: queryArg})
+      return this.member({alias: queryArg}) || this.member({roomAlias: queryArg}) || this.member({name: queryArg})
```
here `alias` & `roomAlias` is a bit confused.

consider to rename to `contactAlias` & `roomAlias` to declare them more clearly.

zixia commented on src/room.ts:
```diff
@@ -423,9 +430,9 @@ export class Room extends EventEmitter implements Sayable {
-      remark:   'remarkMap',
-      display:  'displayMap'
+      name:       'nameMap',
+      alias:      'aliasMap',
```
also use `contactAlias` & `roomAlias` at here.

lijiarui commented:
I guess Coverage decrease because I deleted `const name2 = r.name(contact2)` in `room.spec.ts`, because it is a wrong function caused by me in the first commit.

I'm sorry in my first commit I add a wrong function `Room.name(contact)`, but after read your request, I changed function `name()` into `alias()`.

zixia commented on src/room.ts:
```diff
@@ -353,15 +347,15 @@ export class Room extends EventEmitter implements Sayable {
   public alias(contact: Contact): string {
-    if (!this.obj || !this.obj.nameMap) {
+    if (!this.obj) {
       return ''
```
should `return null` be better?

zixia commented on src/room.ts:
```diff
@@ -353,15 +347,15 @@ export class Room extends EventEmitter implements Sayable {
       return ''
     }
-    return this.obj.roomAliasMap[contact.id] || this.obj.nameMap[contact.id]
+    return this.obj.aliasMap[contact.id]
```
if no `aliasMap[contact.id]`, here'll return a `undefined`, which is not expected?

lijiarui replied:
if no `aliasMap[contact.id]` here, it should return undefiend. Actually, I have deal alias in aliasMap as follows:
```ts
private parseMap(memberList: RoomRawMember[], parseContent: NameType): Map<string, string> {
    const mapList: Map<string, string> = new Map<string, string>()
    if (memberList && memberList.map) {
      memberList.forEach(member => {
        let tmpName: string
        let contact = Contact.load(member.UserName)
        switch (parseContent) {
          case 'name':
            tmpName = contact.alias() || contact.name()
            break
          case 'alias':
            tmpName = member.DisplayName || contact.name()
            break
          default:
            throw new Error('parseMap failed, member not found')
        }
```

if no `DisplayName`, this code `tmpName = member.DisplayName || contact.name()` will get `contact.name()`   

So, no need to add `this.obj.nameMap[contact.id]` here, what do you think about it?

zixia commented on src/contact.ts:
```diff
@@ -222,12 +222,17 @@ export class Contact implements Sayable {
-      query = queryArg
+      if (Object.keys(queryArg)[0] === 'remark') {
+        log.warn('Contact', 'Contact.findAll(remark:%s) DEPRECATED, use Contact.findAll(alias:%s) instead.')
+        query = { alias: queryArg['remark']}
```
queryArg.remark

zixia commented on src/contact.ts:
```diff
@@ -222,12 +222,17 @@ export class Contact implements Sayable {
     let query: ContactQueryFilter
     if (queryArg) {
-      query = queryArg
+      if (Object.keys(queryArg)[0] === 'remark') {
```
`if (queryArg.remark) {`

zixia commented on src/contact.ts:
```diff
@@ -287,51 +292,63 @@ export class Contact implements Sayable {
+    if (newRemark) {
+      return this.alias(newRemark)
+    } else {
+      return this.alias()
```
when `newRemark = null`, will run code here, which is not expected.

zixia commented on src/room.ts:
```diff
@@ -206,10 +204,9 @@ export class Room extends EventEmitter implements Sayable {
-    const remarkMap   = this.parseMap(rawObj.MemberList, 'remark')
-    const displayMap  = this.parseMap(rawObj.MemberList, 'display')
+    const memberList = this.parseMemberList(rawObj.MemberList)
+    const nameMap    = this.parseMap(rawObj.MemberList, 'nick')
```
why `nameMap` use arg name with `nick` here?

should be consistent.

zixia commented on src/room.ts:
```diff
@@ -238,21 +234,20 @@ export class Room extends EventEmitter implements Sayable {
         let contact = Contact.load(member.UserName)
         switch (parseContent) {
           case 'nick':
```
confusing names & logic here...

lijiarui commented on src/room.ts:
```diff
@@ -25,12 +25,11 @@ type RoomObj = {
 }
 
-type NameType = 'name' | 'alias' | 'roomAlias'
+type NameType = 'nick' | 'alias'
```
sorry, my fault..

lijiarui commented on src/contact.ts:
```diff
@@ -227,9 +227,9 @@ export class Contact implements Sayable {
     let query: ContactQueryFilter
     if (queryArg) {
-      if (queryArg[0] === 'remark') {
+      if (Object.keys(queryArg)[0] === 'remark') {
```
I tried it before, but it  isn't compiled successfully:
```
 public static async findAll(queryArg?: ContactQueryFilter | {remark: string | RegExp}): Promise<Contact[]> {
    let query: ContactQueryFilter
    if (queryArg) {
      if (queryArg.remark) {
        query = { alias: queryArg.remark}
      }
```
 error throw:
```
src/contact.ts(230,20): error TS2339: Property 'remark' does not exist on type 'ContactQueryFilter | { remark: string | RegExp; }'.
  Property 'remark' does not exist on type 'ContactQueryFilter'.
src/contact.ts(231,35): error TS2339: Property 'remark' does not exist on type 'ContactQueryFilter | { remark: string | RegExp; }'.
  Property 'remark' does not exist on type 'ContactQueryFilter'.
```

zixia replied:
ok

lijiarui commented on src/contact.ts:
```diff
@@ -227,9 +227,9 @@ export class Contact implements Sayable {
+      if (Object.keys(queryArg)[0] === 'remark') {
         log.warn('Contact', 'Contact.findAll(remark:%s) DEPRECATED, use Contact.findAll(alias:%s) instead.')
-        query = { alias: queryArg[1]}
+        query = { alias: queryArg['remark']}
```
it will throw error

zixia replied:
ok

zixia commented on src/room.ts:
```diff
@@ -350,11 +345,17 @@ export class Room extends EventEmitter implements Sayable {
+      return null
     }
-    return this.obj.displayMap[contact.id] || this.obj.nickMap[contact.id]
+    return this.obj.aliasMap[contact.id]
```
return this.obj.aliasMap[contact.id] || null

zixia commented on src/contact.ts:
```diff
@@ -222,12 +222,17 @@ export class Contact implements Sayable {
+   * find contact by `name` or `alias`
    */
-  public static async findAll(queryArg?: ContactQueryFilter): Promise<Contact[]> {
+  public static async findAll(queryArg?: ContactQueryFilter | {remark: string | RegExp}): Promise<Contact[]> {
```
move `remark: string | RegExp` to `ContactQueryFilter` with a comment noticing _DEPRECATED_

then you could use `.remark` again

lijiarui replied:
> there's no reason to keep remark here anymore. remove it.

you just told me to remove remark from ContactQueryFilter

zixia replied:
Ok, let me tell u one more:

Next time, if u still only know what instead of knowing why, and make stupid argue like this, then kick your ass 10 times by yourself.
